File: 2025/152. Maximum Product Subarray/solution.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        max_neg = max_pos = max_product = nums[0]
        for n in nums[1:]:
            prev_pos = max_pos
            prev_neg = max_neg
            max_pos = max(n, prev_pos * n, prev_neg * n)
            max_neg = min(n, prev_pos * n, prev_neg * n)
            max_product = max(max_pos, max_neg, max_product)
        return max_product

File: 2025/152. Maximum Product Subarray/Solution.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0)
            return 0;
        int res = nums[0], max = nums[0], min = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int n = nums[i];
            int prevMax = max, prevMin = min;
            max = Math.max(n, Math.max(prevMax * n, prevMin * n));
            min = Math.min(n, Math.min(prevMax * n, prevMin * n));
            res = Math.max(Math.max(max, min), res);
        }
        return res;
    }
}